pymongo/asynchronous/mongo_client.py

```diff
@@ -820,87 +822,186 @@ def __init__(
+            # it must be a URI,
+            # https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names
+            if "/" in entity:
+                # Determine connection timeout from kwargs.
+                timeout = keyword_opts.get("connecttimeoutms")
+                if timeout is not None:
+                    timeout = common.validate_timeout_or_none_or_zero(
+                        keyword_opts.cased_key("connecttimeoutms"), timeout
+                    )
+                res = await uri_parser._parse_srv(
+                    entity,
+                    self._port,
+                    validate=True,
+                    warn=True,
+                    normalize=False,
+                    connect_timeout=timeout,
+                    srv_service_name=srv_service_name,
+                    srv_max_hosts=srv_max_hosts,
+                )
+                seeds.update(res["nodelist"])
+                opts = res["options"]
+            else:
+                seeds.update(split_hosts(entity, self._port))
+
+            if not seeds:
+                raise ConfigurationError("need to specify at least one host")
+
+            for hostname in [node[0] for node in seeds]:
+                if _detect_external_db(hostname):
+                    break
+
+            # Add options with named keyword arguments to the parsed kwarg options.
+            tz_aware = keyword_opts["tz_aware"]
+            connect = keyword_opts["connect"]
+            if tz_aware is None:
+                tz_aware = opts.get("tz_aware", False)
+            if connect is None:
+                # Default to connect=True unless on a FaaS system, which might use fork.
+                from pymongo.pool_options import _is_faas
+
+                connect = opts.get("connect", not _is_faas())
+            keyword_opts["tz_aware"] = tz_aware
+            keyword_opts["connect"] = connect
+
+            opts = self._validate_kwargs_and_update_opts(keyword_opts, opts)
+
+            if srv_service_name is None:
+                srv_service_name = opts.get("srvServiceName", common.SRV_SERVICE_NAME)
+
+            srv_max_hosts = srv_max_hosts or opts.get("srvmaxhosts")
+            opts = self._normalize_and_validate_options(opts, seeds)
+
+            # Username and password passed as kwargs override user info in URI.
+            username = opts.get("username", self._resolve_srv_info["username"])
+            password = opts.get("password", self._resolve_srv_info["password"])
+            self._options = ClientOptions(
+                username, password, self._resolve_srv_info["dbase"], opts, _IS_SYNC
+            )
+
+            self._init_based_on_options(seeds, srv_max_hosts, srv_service_name)
+
+    def _init_based_on_options(
+        self, seeds: Collection[tuple[str, int]], srv_max_hosts: Any, srv_service_name: Any
+    ) -> None:
+        self._event_listeners = self._options.pool_options._event_listeners
+        super().__init__(
```

Comment:
I don't think we can call `super().__init__` outside of `__init__`. That would mean we have undefined behavior in between. Like attempting to access `client.codec_options` will fail until the client becomes "connected".